src/fiber/fiber.mli

```
@@ -322,6 +322,21 @@ module Stream : sig
 end
 with type 'a fiber := 'a t
 
+module Pool : sig
+  type t
+
+  type 'a fiber
+
+  val create : unit -> t
+
+  val task : t -> f:(unit -> unit fiber) -> (unit, [ `Stopped ]) result fiber
```

`Stream.Out.write` raises an error rather than return a `result`. Why the difference between the two functions?

---

A stream is usually owned by whoever is responsible for writing to it. While a pool can be shared by many fibers and hence it is harder to guarantee that we aren't writing after stop. The result makes the failure more obvious. This is how I though about it, but I realize it still a matter of taste.

---

If we turn this into the exception, I expect that every call site will need to turn into:

```ocaml
if Pool.stopped pool then
  Fiber.return ()
else 
  Fiber.task pool ~f
```

And that seems intrusive.

---

When do you stop the pool?

---

In the context of RPC, I would expect the RPC server to close the pool when it is shutdown.

---

Ack. Also, there is still an open question: when would we submit a task after the pool is shutdown?

---

Suppose we've requested the shutdown of RPC while the build is still running. The build might still fire some event to RPC even though it's too late to send it.

---

I think I understand looking at the `Fix shutdown` PR. As soon as we get a shutdown request, we immediately stop the RPC. And so here getting `Stopped` means: "we are shuting down".

Looking at `Stream.Out.write` and `Pool.task`, something still feels uncomfortable. These two functions are very similar, but we are choosing a different API based on how we are currently using them in the code base. Fiber feels like a more general library, and I think we should evolve its API more like we would evolve a standalone library, rather than just a part of Dune. So here I would make the same choice for both these functions, given that they serve a similar purpose. i.e.: sending some message to some other fiber.

If the `if Pool.stopped pool then ...` pattern is annoying to write, you could wrap it in a function in the RPC interface, and we could argue there that ignoring the case where the pool is stopped is Ok because it only happens when we are shutting down.

---

Sure, though this also requires adding `val stopped` (or `val running`).

---

Alright this is done.